fdbclient/GlobalConfig.actor.cpp

```diff
@@ -153,95 +153,29 @@ void GlobalConfig::erase(KeyRangeRef range) {
-				    wait(tr->get(Key("\xff\x02/fdbClientInfo/client_txn_sample_rate/"_sr)));
-				state Optional<Value> sizeLimit =
-				    wait(tr->get(Key("\xff\x02/fdbClientInfo/client_txn_size_limit/"_sr)));
-
-				tr->setOption(FDBTransactionOptions::SPECIAL_KEY_SPACE_ENABLE_WRITES);
-				// The value doesn't matter too much, as long as the key is set.
-				tr->set(migratedKey.contents(), "1"_sr);
-				if (sampleRate.present()) {
-					const double sampleRateDbl =
-					    BinaryReader::fromStringRef<double>(sampleRate.get().contents(), Unversioned());
-					Tuple rate = Tuple().appendDouble(sampleRateDbl);
-					tr->set(GlobalConfig::prefixedKey(fdbClientInfoTxnSampleRate), rate.pack());
-				}
-				if (sizeLimit.present()) {
-					const int64_t sizeLimitInt =
-					    BinaryReader::fromStringRef<int64_t>(sizeLimit.get().contents(), Unversioned());
-					Tuple size = Tuple().append(sizeLimitInt);
-					tr->set(GlobalConfig::prefixedKey(fdbClientInfoTxnSizeLimit), size.pack());
-				}
-
-				wait(tr->commit());
-				break;
-			} catch (Error& e) {
-				// If multiple fdbserver processes are started at once, they will all
-				// attempt this migration at the same time, sometimes resulting in
-				// aborts due to conflicts. Purposefully avoid retrying, making this
-				// migration best-effort.
-				TraceEvent(SevInfo, "GlobalConfig_RetryableMigrationError").errorUnsuppressed(e).suppressFor(1.0);
-				wait(tr->onError(e));
-				tr.clear();
-				// tr is cleared, so it won't backoff properly. Use custom backoff logic here.
-				wait(backoff.onError());
-			}
-		}
-	} catch (Error& e) {
-		// Catch non-retryable errors (and do nothing).
-		TraceEvent(SevWarnAlways, "GlobalConfig_MigrationError").error(e);
-	}
-	return Void();
-}
-
 // Updates local copy of global configuration by reading the entire key-range
-// from storage.
-ACTOR Future<Void> GlobalConfig::refresh(GlobalConfig* self) {
-	// TraceEvent trace(SevInfo, "GlobalConfig_Refresh");
+// from storage (proxied through the GrvProxies).
+ACTOR Future<Void> GlobalConfig::refresh(GlobalConfig* self, Version lastKnown) {
+	// TraceEvent trace(SevInfo, "GlobalConfigRefresh");
 	self->erase(KeyRangeRef(""_sr, "\xff"_sr));
 
 	state Backoff backoff;
-
-	state Reference<ReadYourWritesTransaction> tr;
 	loop {
 		try {
-			tr = makeReference<ReadYourWritesTransaction>(Database(Reference<DatabaseContext>::addRef(self->cx)));
-			tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
-			RangeResult result = wait(tr->getRange(globalConfigDataKeys, CLIENT_KNOBS->TOO_MANY));
-			for (const auto& kv : result) {
+			GlobalConfigRefreshReply reply =
+			    wait(timeoutError(basicLoadBalance(self->cx->getGrvProxies(UseProvisionalProxies::False),
+			                                       &GrvProxyInterface::refreshGlobalConfig,
+			                                       GlobalConfigRefreshRequest{ lastKnown }),
+			                      CLIENT_KNOBS->GLOBAL_CONFIG_REFRESH_TIMEOUT));
```

Comment:
It's probably not strictly necessary to time this out, though I suppose one benefit of doing so is that if one proxy has trouble reading this state for some reason, then you will try someone else eventually. The downside of timing it out is that other types of problems could cause you to continually time out even if waiting a little longer would suffice.

A good balance may be to keep the timeout but extend it, say to something like 10s.